classes/plugins/PluginRegistry.php

```diff
@@ -228,93 +157,112 @@ public static function loadPlugin($category, $pathName, $mainContextId = null)
+    /**
+     * Attempts to retrieve plugins from the database.
+     */
+    private static function _loadFromDatabase(string $category, ?int $mainContextId = null): array
+    {
+        $plugins = [];
+        $categoryDir = static::PLUGINS_PREFIX . $category;
+        $products = Application::get()->getEnabledProducts("plugins.{$category}", $mainContextId);
+        foreach ($products as $product) {
+            $name = $product->getProduct();
+            if ($plugin = static::_instantiatePlugin($category, $name, $product->getProductClassname())) {
+                $plugins[$plugin->getSeq()]["{$categoryDir}/{$name}"] = $plugin;
             }
+        }
+        return $plugins;
+    }
 
-            // Try the well-known plug-in class name next (deprecated; pre-namespacing).
-            // (DEPRECATED as of 3.4.0.)
-            $pluginClassName = ucfirst($file) . ucfirst($category) . 'Plugin';
-            $pluginClassFile = $pluginClassName . '.inc.php';
-            if (file_exists("${pluginPath}/${pluginClassFile}")) {
-                // Try to instantiate the plug-in class.
-                $pluginPackage = 'plugins.' . $category . '.' . $file;
-                $plugin = instantiate($pluginPackage . '.' . $pluginClassName, $pluginClassName, $pluginPackage, 'register');
-                assert(is_a($plugin, $classToCheck ?: '\PKP\plugins\Plugin'));
-                return $plugin;
+    /**
+     * Get all plug-ins from disk without querying the database, used during installation.
+     */
+    private static function _loadFromDisk(string $category): array
+    {
+        $categoryDir = static::PLUGINS_PREFIX . $category;
+        if (!is_dir($categoryDir)) {
+            return [];
+        }
+        $plugins = [];
+        foreach (new FilesystemIterator($categoryDir) as $path) {
+            if (!$path->isDir()) {
+                continue;
             }
+            $pluginName = $path->getFilename();
+            if ($plugin = static::_instantiatePlugin($category, $pluginName)) {
+                $plugins[$plugin->getSeq()]["{$categoryDir}/{$pluginName}"] = $plugin;
+            }
+        }
+        return $plugins;
+    }
+
+    /**
+     * Instantiate a plugin.
+     * @deprecated 3.4.0 Old way to instantiate a plugin
+     */
+    private static function _deprecatedInstantiatePlugin(string $category, string $pluginName): ?Plugin
+    {
+        $pluginPath = static::PLUGINS_PREFIX . "${category}/{$pluginName}";
+        // Try the plug-in wrapper for backwards compatibility.
+        $pluginWrapper = "${pluginPath}/index.php";
+        if (file_exists($pluginWrapper)) {
+            return include $pluginWrapper;
+        }
+
+        // Try the well-known plug-in class name next (with and without ".inc.php")
+        $pluginClassName = ucfirst($pluginName) . ucfirst($category) . 'Plugin';
+        if (Arr::first(['.inc.php', '.php'], fn (string $suffix) => file_exists("${pluginPath}/{$pluginClassName}{$suffix}"))) {
```

Comment:
Here I've added an extra check to look for `.php` and `.inc.php`, which is already supported by the `instantiate()` function below.

Reply:
This behaviour was marked deprecated but that warning has been removed... https://github.com/pkp/pkp-lib/blob/72565d9cd4fb2dc6a8d3a82988463b6cf3665c2e/classes/plugins/PluginRegistry.php#L303